src/photon/photon_algorithm.c

```diff
@@ -254,14 +334,19 @@ void handle_task_submitted(local_scheduler_state *state,
    * cannot assign the task to a worker immediately, we either queue the task in
    * the local task queue or we pass the task to the global scheduler. For now,
    * we pass the task along to the global scheduler if there is one. */
-  if ((utarray_len(algorithm_state->available_workers) > 0) &&
-      can_run(algorithm_state, spec)) {
-    run_task_immediately(state, algorithm_state, spec, false);
-  } else if (state->db == NULL || !state->global_scheduler_exists) {
-    queue_task_locally(state, algorithm_state, spec, false);
+  if (can_run(algorithm_state, spec) &&
+      (utarray_len(algorithm_state->available_workers) > 0)) {
+    /* Dependencies are ready and there is an available worker, so dispatch the
+     * task. */
+    queue_dispatch_task(state, algorithm_state, spec);
   } else {
-    give_task_to_global_scheduler(state, algorithm_state, spec, false);
+    /* Give the task to the global scheduler to schedule, if it exists. */
+    give_task_to_global_scheduler(state, algorithm_state, spec);
```

Comment:
It's a special case if there's no global scheduler, used only for testing purposes. I wanted to do the if statement inside of the `give_task_to_global_scheduler` function to hide away the details of passing it to the "global scheduler".